manta-pay/src/crypto/hash/poseidon/parameter_generation/matrix.rs

```diff
@@ -0,0 +1,900 @@
+            current = current.eliminate(column, &mut shadow_matrix)?;
+            result.push(current.0.remove(0));
+            shadow_result.push(shadow_matrix.0.remove(0));
+        }
+        result.push(current.0.take_first());
+        shadow_result.push(shadow_matrix.0.take_first());
+        *shadow = Self(Matrix(shadow_result));
+        Some(Self(Matrix(result)))
+    }
+}
+
+impl<F> AsRef<Matrix<F>> for SquareMatrix<F>
+where
+    F: Field,
+{
+    fn as_ref(&self) -> &Matrix<F> {
+        &self.0
+    }
+}
+
+impl<F> Deref for SquareMatrix<F>
+where
+    F: Field,
+{
+    type Target = Matrix<F>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl<F> PartialEq<Matrix<F>> for SquareMatrix<F>
+where
+    F: Field + PartialEq,
+{
+    fn eq(&self, other: &Matrix<F>) -> bool {
+        self.0.eq(other)
+    }
+}
+
+impl<F> MatrixOperations for SquareMatrix<F>
+where
+    F: Field,
+{
+    type Scalar = F;
+
+    fn eliminate(&self, column: usize, shadow: &mut Self) -> Option<Self>
+    where
+        Self::Scalar: Clone,
+    {
+        self.0.eliminate(column, &mut shadow.0).map(Self)
+    }
+
+    fn identity(n: usize) -> Self {
+        Self(Matrix::identity(n))
+    }
+
+    fn matmul(&self, other: &Self) -> Option<Self>
+    where
+        Self::Scalar: Clone,
+    {
+        self.0.matmul(&other.0).map(Self)
+    }
+
+    fn mul_by_scalar(&self, scalar: Self::Scalar) -> Self {
+        Self(self.0.mul_by_scalar(scalar))
+    }
+
+    fn to_row_major(self) -> Vec<F> {
+        self.0.to_row_major()
+    }
+
+    fn transpose(self) -> Self {
+        Self(self.0.transpose())
+    }
+}
+
+/// Inner product of two vectors.
+pub fn inner_product<F>(a: &[F], b: &[F]) -> F
+where
+    F: Field,
+{
+    a.iter()
+        .zip(b)
+        .fold(F::zero(), |acc, (v1, v2)| F::add(&acc, &F::mul(v1, v2)))
+}
+
+/// Elementwise addition (i.e., out_i = a_i + b_i).
+pub fn vec_add<F>(a: &[F], b: &[F]) -> Vec<F>
+where
+    F: Field,
+{
+    a.iter().zip(b).map(|(a, b)| F::add(a, b)).collect()
+}
+
+/// Elementwise subtraction (i.e., out_i = a_i - b_i).
+pub fn vec_sub<F>(a: &[F], b: &[F]) -> Vec<F>
+where
+    F: Field,
+{
+    a.iter().zip(b.iter()).map(|(a, b)| F::sub(a, b)).collect()
+}
+
+/// Elementwisely multiplies a vector `v` with `scalar`.
+pub fn scalar_vec_mul<F>(scalar: &F, v: &[F]) -> Vec<F>
+where
+    F: Field,
+{
+    v.iter().map(|val| F::mul(scalar, val)).collect()
+}
+
+/// Eliminates `row` with `factor` multiplying `pivot`.
+fn eliminate_row<F>(row: &[F], factor: &F, pivot: &[F]) -> Vec<F>
+where
+    F: Field,
+{
+    vec_sub(row, &scalar_vec_mul(factor, pivot))
+}
+
+/// Returns kronecker delta.
+pub fn kronecker_delta<F>(i: usize, j: usize) -> F
+where
+    F: Field,
+{
+    if i == j {
+        F::one()
+    } else {
+        F::zero()
+    }
+}
+
+/// Checks whether `elem` equals zero.
+pub fn equal_zero<F>(elem: &F) -> bool
+where
+    F: Field,
+{
+    F::eq(elem, &F::zero())
+}

```

Comment:
Is this function used? If it is, then add it to the `Field` trait and call it `is_zero`.